codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -375,13 +381,51 @@ public void write(int allowedBytes) {
                         writePromise = ctx.newPromise();
                         writePromise.addListener(this);
                     }
+                    if (toWrite instanceof SlicedByteBuf && data instanceof CompositeByteBuf) {
```

Comment:
If the underlying buffer is composite then either were writing all of it or were writing a slice. If it's a slice (this check) then we want to free the underlying buffers associated with the slice.